### DISCUSSION TITLE
Self-Hosted JSPM

### DISCUSSION THREAD

**Discussion by bennypowers** [jspm/project#205] [Q&A]
We'd like to host our own JSPM CDN. How likely is that to ever happen?

**Comment by guybedford** [comment]
It's a certainty if you are in a position for your company to pay for a license to do that. I would happily put the necessary time in to set it up for that, with the work directly going to the JSPM foundation as a sponsorship for the free CDN.

**Reply by bennypowers** [reply]
What kind of license would it be?

**Reply by guybedford** [reply]
It would need to be worked out - it could be a one-off or subscription license. There would probably be at least 20-40 hours of work on my side to get something together that can be supported, so any fee should at least cover 2x that from a consulting time perspective to make it worthwhile. If you'd like to discuss further I can certainly put a proposal together.

**Comment by trusktr** [comment]
I like the idea of self-hosting.

@guybedford you could perhaps make the server open source with optional paid support plan, and also make a website that has an way to spin up private non-self-hosted registries for a recurring cost with custom domains. This could include the importmap generator, so that it can generate import maps using the URL of the private registry, with fallback to the JSPM registry (kinda like private NPM registries that fallback to the main one).

This could be interesting because then people making libraries could have a nice way to instruct users to install their library, and it'll all come from their domains. Giving a nice unified identity for the source of the library code.

Once this is in place, you could perhaps allow logged-in users to register their registries, and make them the canonical for their library, so that the main JSPM CDN and importmap generator can generate maps that include the domains of all the libraries, making it kinda like a "decentralized" package registry, while a centralized tool can wire it all up together.

**Reply by trusktr** [reply]
I think that open licensing would be a good strategy here compared to trying to get licenses for a closed source setup. The [generator has issues](https://github.com/jspm/project/issues?q=is%3Aissue+is%3Aopen+importmap+generator), the [CDN has issues](https://github.com/jspm/project/issues?q=is%3Aissue+is%3Aopen+cdn), and right now only one person is working on them is (basically @guybedford).

If everything were opened up, with permissive license, and people were able to contribute in *every* part of all of JSPM, that could be beneficial for making the tech much more stable and useful, and taking it to new levels, and really changing the open-source-minded JS ecosystem.

Then, with more adoption (people not avoiding the tool because they face issues), building a business on top of the bigger ecosytem could be more successful.

**Reply by guybedford** [reply]
@trusktr all the issues you have posted are with open source components, therefore I struggle to see how your argument for contributions holds.

Building a business or using a business / open source model seems the right way to approach it overall. At the very least partnering with businesses that can help sustain the project.

Partnering on self-hosting is exactly the hope to bring in income, which wouldn't happen if it was already open source. If there was a valid sustainable development model, then I would certainly open source the CDN. But until that's worked out, or until we can work out how to run this project sustainably in the long term (the CDN costs $3000 per month to run), these answers aren't obvious.
